### PR TITLE
Run driver with user context

### LINKED ISSUE #100164: [CI] EsqlSpecIT test {ints.MultipleWarnings} failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/juee6bgkgo6na/tests/:x-pack:plugin:esql:qa:server:multi-node:javaRestTest/org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT/test%20%7Bints.MultipleWarnings%7D
**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:qa:server:multi-node:javaRestTest' --tests "org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT" -Dtests.method="test {ints.MultipleWarnings}" -Dtests.seed=CAB3ECC97601B05C -Dtests.locale=zh-HK -Dtests.timezone=Asia/Brunei -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT&tests.test=test%20%7Bints.MultipleWarnings%7D
**Failure excerpt:**
```
java.lang.AssertionError: Expected a list containing
 0: "Line 1:76: evaluation of [to_int(x)] failed, treating result as null. Only first 20 failures recorded."
 1: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100010000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100680000000] out of [integer] range"
 2: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100020000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100690000000] out of [integer] range"
 3: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100030000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100700000000] out of [integer] range"
 4: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100040000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100710000000] out of [integer] range"
 5: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100050000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100720000000] out of [integer] range"
 6: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100060000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100730000000] out of [integer] range"
 7: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100070000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100740000000] out of [integer] range"
 8: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100080000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100750000000] out of [integer] range"
 9: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100090000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100760000000] out of [integer] range"
10: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100100000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100770000000] out of [integer] range"
11: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100110000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100780000000] out of [integer] range"
12: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100120000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100790000000] out of [integer] range"
13: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100130000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100800000000] out of [integer] range"
14: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100140000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100810000000] out of [integer] range"
15: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100150000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100820000000] out of [integer] range"
16: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100160000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100830000000] out of [integer] range"
17: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100170000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100840000000] out of [integer] range"
18: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100180000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100850000000] out of [integer] range"
19: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100190000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100860000000] out of [integer] range"
20: expected "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100200000000] out of [integer] range" but was "Line 1:76: org.elasticsearch.xpack.ql.QlIllegalArgumentException: [100870000000] out of [integer] range"

  at __randomizedtesting.SeedInfo.seed([CAB3ECC97601B05C:42E7D313D8FDDDA4]:0)
  at org.elasticsearch.test.MapMatcher.assertMap(MapMatcher.java:82)
  at org.elasticsearch.test.MapMatcher.assertMap(MapMatcher.java:65)
  at org.elasticsearch.xpack.esql.qa.rest.RestEsqlTestCase.performRequest(RestEsqlTestCase.java:498)
  at org.elasticsearch.xpack.esql.qa.rest.RestEsqlTestCase.runEsql(RestEsqlTestCase.java:447)
  at org.elasticsearch.xpack.esql.qa.rest.EsqlSpecTestCase.doTest(EsqlSpecTestCase.java:102)
  at org.elasticsearch.xpack.esql.qa.rest.EsqlSpecTestCase.test(EsqlSpecTestCase.java:94)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by dnhatn**
Today, we have a hierarchy of tasks in ESQL designed to leverage the task framework for reporting status and cancellation.

```mermaid
flowchart
 RESTLayer -->|  EsqlQueryRequest indices:data/read/esql  | ComputeService
 ComputeService -->| DriverRequest indices:data/read/esql/compute | Driver
 ComputeService -->| DataNodeRequest indices:data/read/esql/data | DataNode
 DataNode -->| DriverRequest indices:data/read/esql/compute | Driver
 Driver -->| LookupRequest indices:data/read/esql/lookup | EnrichLookupService
```

The primary issue here is that `DriverRequest` is neither `IndicesRequest` nor `CompositeIndicesRequest`. Consequently, the Driver is executed within the context of the system user, leading to access indices with the system user.

To address this issue, this PR makes `DriverRequest` a `CompositeIndicesRequest` and ensures that the Driver executes within the user's context. With this fix we can now properly capture the response headers when a Driver is yielded and rescheduled (see https://github.com/elastic/elasticsearch/issues/100164#issuecomment-1753772700).

Relates #100707
Relates #99646

Relates #99926
Closes #100164

### CODE REVIEW

**Comment by jakelandis** [comment]
There seems to be a bug in our docs introduced here: https://github.com/elastic/elasticsearch/pull/85877 `manage_enrich` is not role, it is privilege [1]. I will get that corrected. 

> Regarding the privilege requirements for users to use ESQL enrich, it's currently unclear whether they need the enrich_user role or if having just the monitor_enrich role is sufficient.

I don't think we ever mandate the use of the roles we ship by default, they are generally helpful but more for UX concerns than for security concerns. A user can always add the needed privileges to a custom role.  

[1]
We have some naming problems as evident by some confusion in that PR. To help clarify: Access is granted on action names (i.e. cluster:monitor/xpack/enrich/foo), multiple action names are rolled up into privileges (i.e. monitor_enrich -> cluster:monitor/xpack/enrich/* ), multiple privileges are added to a roles and we ship some roles by default (i.e. enrich_user -> monitor_enrich). Users have one or more roles (i.e. bob -> enrich_user) and those roles can be ones we ship or ones created by the end user (note - end users can not currently create custom roles in serverless) . So bob (user) access looks like enrich_user (role) -> monitor_enrich (priv) -> cluster:monitor/xpack/enrich/foo (action).

**Comment by dnhatn** [comment]
@jakelandis Apologies for any confusion; I used the wrong terminology. Let me clarify the situation: currently, we have two cluster privileges: `monitor_enrich` and `manage_enrich`, along with the `enrich_user` role that has the `manage_enrich` privilege and access to the `.enrich-*` indices. To resolve the enrich policy, it currently requires the `monitor_enrich` cluster privilege, which is a subset of the `manage_enrich` cluster privilege, as well as the `enrich_user` role. At the moment, there is no requirement for users to have access to the `.enrich-*` indices to use enrich in ESQL. Should we consider requiring users to have access to the `.enrich-*` indices to use enrich in ESQL? I initially thought Yang suggested that the `monitor_enrich` privilege should be sufficient, but this does not seem to align with our principle. I've removed this question from the PR's description.

Another issue (not related to this PR) is that we should have both roles: `enrich_admin` for manage and `enrich_user` for read-only. However, the `enrich_user` currently has a manage privilege.

**Comment by jakelandis** [comment]
> Should we consider requiring users to have access to the .enrich-* indices to use enrich in ESQL?

I don't think so. Users should never need direct access to system indices to perform normal operations. 

Also, I am not sure I can properly review this PR as I am not very familiar with much of this area. However, happy to consult w.r.t. getting the right set of user/privs in place.

**Comment by jakelandis** [comment]
> Another issue (not related to this PR) is that we should have both roles: enrich_admin for manage and enrich_user for read-only. However, the enrich_user currently has a manage privilege.

I don't think that `enrich_admin` should be necessary. We can not passively lower permissions and enrich_user and (not sure if this implied or not) we shouldn't give direct write access to the .enrich-* indices. The enrich user purpose (today) is to setup and/or troubleshoot enrich configuration/execution so it is already admin like.

**Comment by dnhatn** [comment]
> I don't think so. Users should never need direct access to system indices to perform normal operations.

Thanks @jakelandis. This is the feedback that I need. I will open a separate PR for this, which I think is much contained and related to the security.

**Comment by dnhatn** [comment]
@jakelandis Thanks for looking. I've opened https://github.com/elastic/elasticsearch/pull/100856, which includes only the security related change.

**Comment by dnhatn** [comment]
> [PowTests » testEvaluate {TestCase=pow(integer, double)}](https://gradle-enterprise.elastic.co/s/k6zfcys4aokxy/tests/:x-pack:plugin:esql:test/org.elasticsearch.xpack.esql.expression.function.scalar.math.PowTests/testEvaluate%20%7BTestCase=pow(integer,%20double)%7D)

run elasticsearch-ci/part-2

**Comment by dnhatn** [comment]
I'm marking this PR as a blocker for 8.11 since ESQL enrich might not work with security enabled.

**Comment by dnhatn** [comment]
@nik9000, it would be great if you could review this tomorrow when you're back.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java [hunk 1]
```diff
@@ -354,4 +375,22 @@ private DriverStatus updateStatus(DriverStatus.Status status) {
             activeOperators.stream().map(o -> new DriverStatus.OperatorStatus(o.toString(), o.status())).toList()
         );
     }
+
+    private static class StoredThreadContext {
```
Comment (jakelandis) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java:379]: the usage of this looks similar to `org.elasticsearch.action.support.ContextPreservingActionListener` , could that work here instead ? Dealing with thread context is tricky and the try-with-resource provides some nice guarantees. (the code is not obvious but when the try with resources in ContextPreservingActionListener terminates will restore the passed in context)

Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java]: @jakelandis I tried using the ContextPreservingActionListener before, but it didn't work as expected. I believe the issue was that I reserved some context twice. I've made this work with 0bf6c22. Thanks for pushing on this :).

**Comment by dnhatn** [comment]
@jakelandis This PR includes 4 important changes; the rest are mostly plumbing matters, which you can skip entirely.

https://github.com/elastic/elasticsearch/pull/100724/files#diff-d88ab9ce817937201c29201c2a835be4e7f55e78524092eb718676b2dbeb810cR37

https://github.com/elastic/elasticsearch/pull/100724/files#diff-cafe8ecff527a94253e2305e3724d3e0416373c45a86e568bbef44ed54667539R270

https://github.com/elastic/elasticsearch/pull/100724/files#diff-d88ab9ce817937201c29201c2a835be4e7f55e78524092eb718676b2dbeb810cR62

https://github.com/elastic/elasticsearch/pull/100724/files#diff-e7a17001be18e2beff7eb7ab8c77ff40d01eb9c95b39bb98362e8b33ab07389aR310

**Review by jakelandis — APPROVED** [review]
The security bits LGTM, but some of the other context/code is beyond my understanding so please have an additional reviewer before merging. 

The security change is as described, (assuming no upstream (from the start method) manipulations of the thread context) the user's context will be used and preserved with response headers after the listener is handled. It is worth noting that even with the change to use `CompositeIndicesRequest` + ignoring the index level checks, the user must still have `read` access to some index, any index somewhere in their in their role. Authz will (correctly) fail the request in that case but it would be nice (but not a blocker) to have some functional tests that ensure the desired behavior in that case (I think it will error differently with enrich vs. without enrich). Additionally it would be nice for functional test for case where the user does not have the monitor_enrich priv (per https://github.com/elastic/elasticsearch/pull/100856)

**Review by nik9000 — APPROVED** [review]
I don't understand the security issue all that well, but I'm happy with the change and the warnings flow. I'd love to get more high level security tests for the actual opening this fixes, but I'm happy with the code either way.

**Comment by dnhatn** [comment]
> I'd love to get more high level security tests for the actual opening this fixes

@nik9000 thank you for the quick review. This fix removes an AwaitsFix in the enrich security IT test: https://github.com/elastic/elasticsearch/pull/100724/files#diff-2a2742f4edb51f606a43b463079785ed6c33fbe9295fa03b04fb5db758d544f6L121

**Pull request merged**